Use Shared IRQ for stm32 ADC driver

mcscholtz opened the pull request:
I added the ability to use a shared IRQ for the stm32 ADC driver to allow devices with just a single ADC IRQ to use all channels.

Edit: This is my first PR, please let me know if I need to make any changes. Thanks!

Fixes #26874

mcscholtz commented:
Looks like the failures are due to the configurable prescalar change. I will take a look later when I have time

mcscholtz commented:
I was reading the contribution guidelines and I missed a few things, such as the commit signoff. I will fix that soon.

mcscholtz commented:
I fixed the styling issues that caused the compliance checks to fail before. I also veried it with the checkpatch script before pushing it.

mbolivar-nordic commented on drivers/adc/Kconfig.stm32:
```diff
@@ -3,10 +3,21 @@
+	depends on ADC_STM32
+	help
+	  Enable the use of shared interrupts for famalies that only have a single interrupt for all ADC's
+endif
```
Please add a newline to the end of the file.

mcscholtz replied:
The compliance check failed after adding the new line at the end of the file

mbolivar-nordic commented on drivers/adc/adc_stm32.c:
```diff
@@ -227,7 +236,6 @@ struct adc_stm32_data {
 	const struct device *dev;
 	uint16_t *buffer;
 	uint16_t *repeat_buffer;
-
```
We usually try not to include unrelated whitespace changes in commits which change functionality. I would recommend dropping this change or moving it to a separate commit.

mbolivar-nordic commented on drivers/adc/adc_stm32.c:
```diff
@@ -239,12 +247,17 @@ struct adc_stm32_data {
 
 struct adc_stm32_cfg {
 	ADC_TypeDef *base;
+	const uint16_t prescalar;
```
No need for `const` here as the entire structure is `const`.

mbolivar-nordic commented on drivers/adc/adc_stm32.c:
```diff
@@ -642,7 +687,7 @@ static int adc_stm32_check_acq_time(uint16_t acq_time)
 		return 0;
 	}
 
-	LOG_ERR("Conversion time not supportted.");
+	LOG_ERR("Conversion time not supported.");
```
You might want to split this into its own commit also.

mcscholtz replied:
@mbolivar-nordic I can do that, would you recommend to just remove this and the other changes and force update the single comment. Then create new comments for the other changes, like the fix of this typo in this case?

mbolivar-nordic replied:
How people split up commits while rebasing is a personal decision, so I can't really offer you a "recommendation" in that sense. Beyond individual preference, tooling varies: for example, I use [magit](https://magit.vc/) instead of the command line.

However, one simple way to handle this since the PR contains a single commit is to put all its changes back into the working tree:

```
git checkout feature/stm32f407-adc
git reset HEAD~
```

Then you can split changes in a single file into their own commits using `git add -p`, e.g. for this particular comment:

```
git add -p drivers/adc/adc_stm32.c
< follow instructions to just stage this hunk>
git commit
<write a commit comment describing the typo fix>
```

Then keep adding additional commits for the other changes you want to make.

But again this sort of thing is subject to taste, and I would do it in a different way using my own favorite tools.

mbolivar-nordic commented on dts/arm/st/f4/stm32f4.dtsi:
```diff
@@ -442,7 +443,7 @@
 
 		adc1: adc@40012000 {
 			compatible = "st,stm32-adc";
-			reg = <0x40012000 0x400>;
+			reg = <0x40012000 0x50>;
```
Is this really part of the same commit? I think it needs a justification.

mcscholtz replied:
@mbolivar-nordic The ADC1 device currently holds the address space for ADC1, ADC2, ADC3. 

The address for all ADC's starts at 0x40012000 with a length of 0x400. The following is the addresses of the individual ADC's:
ADC1: 0x40012000 length of 0x50
ADC2: 0x40012100 length of 0x50
ADC3: 0x40012200 length of 0x50
Common: 0x40012300 length of 0xC (need to double check the length of this one)

Maybe someone that knows stm32 well can comment on how to handle it.

mbolivar-nordic replied:
OK, then yes, I think this definitely needs a justification and likely a separate commit. 

> Maybe someone that knows stm32 well can comment on how to handle it.

Let's wait to see what the STM32 maintainers say.

erwango replied:
I do agree with @mbolivar-nordic. Please split this in a dedicated commit.

mbolivar-nordic reviewed (changes requested):
Thanks for your first patch! I have some general coding style comments. Please remember to follow the contribution guidelines by squashing or splitting commits as needed, without piling on a series of "fix this, fix that" commits to address review feedback.

@erwango or another STM32 maintainer will need to do the hardware specific review, but the binding change looks fine to me.

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -968,6 +1048,34 @@ static const struct adc_driver_api api_stm32_driver_api = {
+static const struct adc_stm32_cfg adc_stm32_cfg_##index = {		\
+	.base = (ADC_TypeDef *)DT_INST_REG_ADDR(index),			\
+	.prescalar = DT_INST_PROP_OR(index, prescalar, 4),		\
+	.irq_cfg_func = adc_stm32_shared_irq_handler,			\
```
This part of the struct is actually meant to hold the irq init function, not the irq handler.
In current case, it should then be NULL if we don't want to use it.
Alternatively, it could be set to `adc_stm32_cfg_func_isr` and use the `init_irq` flag inside this function to make sure irq init is only done once (which would remove need for the `adc_stm32_init` hack).

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1005,4 +1114,8 @@ static void adc_stm32_cfg_func_##index(void)				\
 	irq_enable(DT_INST_IRQN(index));				\
 }
 
+#ifdef CONFIG_ADC_STM32_SHARED_IRQS
+DT_INST_FOREACH_STATUS_OKAY(STM32_ADC_INIT_SHARED)
+#else
 DT_INST_FOREACH_STATUS_OKAY(STM32_ADC_INIT)
+#endif
```
Alternative otpion is to put the `#ifdef CONFIG_ADC_STM32_SHARED_IRQS` inside `STM32_ADC_INIT` definition, to minimize code changes

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -681,6 +727,49 @@ static void adc_stm32_setup_speed(const struct device *dev, uint8_t id,
+#if defined(CONFIG_SOC_SERIES_STM32F2X) ||	\
+	defined(CONFIG_SOC_SERIES_STM32F4X) ||	\
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+static uint32_t adc_stm32_get_prescalar(uint16_t prescalar)
```
Please put the prescalar changes into a dedicated commit.

erwango commented on drivers/adc/Kconfig.stm32:
```diff
@@ -3,10 +3,21 @@
+	default y
+	depends on ADC_STM32
+	help
+	  Enable the use of shared interrupts for famalies that only have a single interrupt for all ADC's
```
s/famalies/families

erwango reviewed (changes requested):
Thanks for this addition!
Some changes requested otherwise this looks fine.

Then, have you been able to test these changes ?

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -984,7 +1038,16 @@ static const struct adc_stm32_cfg adc_stm32_cfg_##index = {		\
+static const struct soc_gpio_pinctrl adc_pins_##index[] =		\
+	ST_STM32_DT_INST_PINCTRL(index, 0);				\
+									\
+	ADC_STM32_CONFIG_ISR(index)					\
```
Unindent for readability.

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -968,12 +973,61 @@ static const struct adc_driver_api api_stm32_driver_api = {
+	}
+}
+
+#define ADC_STM32_CONFIG_ISR(index)					\
```
I see why you used `ADC_STM32_CONFIG_ISR`, but this is a bit misleading.
Let's use `ADC_STM32_CONFIG` instead of `ADC_STM32_CONFIG_ISR`.

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -968,12 +973,61 @@ static const struct adc_driver_api api_stm32_driver_api = {
+	DT_INST_FOREACH_STATUS_OKAY(HANDLE_IRQS);
+}
+
+static void adc_stm32_cfg_func_isr(void)
```
This funtion is not technically an isr, what about `adc_stm32_irq_init` ?

erwango reviewed (changes requested):
Thanks for the update, few more comments.

mcscholtz commented:
@erwango I made the changes you requested

mcscholtz commented:
@mbolivar-nordic Adding a line at the end of file drivers/adc/Kconfig.stm32 caused the compliance check to fail. Do you want me to remove it again?

mbolivar-nordic commented:
> Adding a line at the end of file drivers/adc/Kconfig.stm32 caused the compliance check to fail. Do you want me to remove it again?

This can be a bit tricky, but there is a difference between "newline at end of file" and "empty line at end of file".

I'm asking for "newline at end of file", which means that the last byte in the file is a newline character. What is in there right now is "empty line at end of file", which is **two** newline characters.

Commit:
drivers: arm: stm32: adc: Added Shared IRQ

Certain stm32 soc's have a single shared irq for all adc channels
on those soc's only a single channel is supported.

Added a Kconfig that enables a shared irq for stm32f2x, stm32f4x,
stm32f7x soc's. The shared IRQ uses a flag to limit the number of
interrupts defined to only 1. A shared irq handler is added which
determines which ADC instance the interrupt is for, it then calls
into the existing interrupt.

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>

mcscholtz commented:
> I'm asking for "newline at end of file", which means that the last byte in the file is a newline character. What is in there right now is "empty line at end of file", which is **two** newline characters.

Okay, thanks for clarifying that. I have removed the extra new line

ABOSTM reviewed (approved):
LGTM

mcscholtz committed:
Merge branch 'main' into feature/stm32f407-adc

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -982,12 +987,61 @@ static const struct adc_driver_api api_stm32_driver_api = {
+
+static void adc_stm32_irq_init(void)
+{
+	if (init_irq) {
```
If we want this piece of code to be executed only once, we have to rely on a global variable, imo.

MaureenHelm reviewed (changes requested):
> @mcscholtz You will have to rebase and make sure that you only have one ADC commit on top of the branch if you continue with this branch.

carlescufi commented:
> > @mcscholtz You will have to rebase and make sure that you only have one ADC commit on top of the branch if you continue with this branch.

@MaureenHelm would you pick this PR up? Seems a shame not to merge it if the author is unresponsive.

mcscholtz commented:
Closing. See https://github.com/zephyrproject-rtos/zephyr/pull/41497
